h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -0,0 +1,251 @@
+    public double[][][] _binvD; // store BinvD for each gam column specified for scoring
+    public double[][] _knots; // store knots location for each gam column
+    public int[] _numKnots;  // store number of knots per gam column
+    String[][] _gamColNames;  // store column names of GAM columns
+    String[][] _gamColNamesCenter;  // gamColNames after de-centering is performed.
+    Key<Frame>[] _gamFrameKeys;
+    Key<Frame>[] _gamFrameKeysCenter;
+
+    /***
+     * This method will take the _train that contains the predictor columns and response columns only and add to it
+     * the following:
+     * 1. For each predictor included in gam_x, expand it out to calculate the f(x) and attach to the frame.
+     * 2. It will calculate the ztranspose that is used to center the gam columns.
+     * 3. It will calculate a penalty matrix used to control the smoothness of GAM.
+     * 
+     * @return
+     */
+    Frame adaptTrain() {
+      int numGamFrame = _parms._gam_X.length;
+      _centerGAM = (numGamFrame > 1) || (_train.numCols() - 1 + numGamFrame) >= 2;
+      _zTranspose = _centerGAM ? GamUtils.allocate3DArray(numGamFrame, _parms, 0) : null;
+      _penalty_mat = _centerGAM ? GamUtils.allocate3DArray(numGamFrame, _parms, 2) :
+              GamUtils.allocate3DArray(numGamFrame, _parms, 1);
+      _binvD = GamUtils.allocate3DArray(numGamFrame, _parms, 3);
+      _numKnots = MemoryManager.malloc4(numGamFrame);
+      _knots = new double[numGamFrame][];
+      _gamColNames = new String[numGamFrame][];
+      _gamColNamesCenter = new String[numGamFrame][];
+      _gamFrameKeys = new Key[numGamFrame];
+      _gamFrameKeysCenter = new Key[numGamFrame];
+
+      addGAM2Train(_parms, _parms.train(), _zTranspose, _penalty_mat, _gamColNames, _gamColNamesCenter,
+              true, _centerGAM, _gamFrameKeys, _gamFrameKeysCenter, _binvD, _numKnots, _knots);
+      return buildGamFrame(numGamFrame, _gamFrameKeys, _train, _parms._response_column); // add gam cols to _train
+    }
+
+    @Override
+    public void computeImpl() {
+      init(true);     //this can change the seed if it was set to -1
+      if (error_count() > 0)   // if something goes wrong, let's throw a fit
+        throw H2OModelBuilderIllegalArgumentException.makeFromBuilder(GAM.this);
+
+      _job.update(0, "Initializing model training");
+
+      buildModel(); // build gam model 
+    }
+
+    public final void buildModel() {
+      GAMModel model = null;
+      DataInfo dinfo = null;
+      Frame newTFrame=null, gamColsFrame=null, gamColsFrameCenter=null;
+      try {
+        _job.update(0, "Adding GAM columns to training dataset...");
+        int trainCols = _train.numCols();  // number of columns before adding GAM columns
+        String[] trainColNames = _train.names(); // column names before adding GAM columns
+        Frame newTrain = rebalance(adaptTrain(), false, _result+".temporary.train"); // add and store gam cols without centering
```

Comment:
newTrain contains predictor variables and basis functions specified in _parms._gam_X without applying zero-sum constraint.